src/coreclr/vm/loaderallocator.cpp

```diff
@@ -919,10 +919,11 @@ OBJECTREF LoaderAllocator::CompareExchangeValueInHandle(LOADERHANDLE handle, OBJ
     if ((((UINT_PTR)handle) & 1) != 0)
     {
         OBJECTREF *ptr = (OBJECTREF *)(((UINT_PTR)handle) - 1);
-        gc.previous = *ptr;
-        if ((*ptr) == gc.compare)
+
+        gc.previous = InterlockedCompareExchangeT(ptr, gc.value, gc.compare);
+        if (gc.previous == gc.compare)
         {
-            SetObjectReference(ptr, gc.value);
+            ErectWriteBarrier(ptr, gc.value);
```

Comment:
I'd like to confirm my understanding here. The handle above was properly updated using `InterlockedCompareExchangeT`, but we need to also update the GC's card tables so need to also call `ErectWriteBarrier`, right?

Reply:
Right

Reply:
does this also need
```
#ifdef _DEBUG
        Thread::ObjectRefAssign(ptr);
#endif
```

Perhaps it is easier to just call `COMInterlocked::CompareExchangeObject(ptr, gc.value, gc.compare)` ?

Reply:
`Thread::ObjectRefAssign(ptr);` is not necessary here. It is only necessary when the assigned OBJECTREF is OBJECTREF on stack. It is not the case here.

`COMInterlocked::CompareExchangeObject` is FCall. It is difficult to call FCalls from guts of the runtime. It requires to suppressing number of contract violations.

Reply:
I guess the need to InterlockExchange references does not come too often, so we do not have a regular, not an FCall helper. 
It is indeed easy to forget about the barrier, which is needed, since this can be a cross-generational assignment.